app/src/main/java/org/kiwix/kiwixmobile/main/AddNoteDialog.java

```diff
@@ -0,0 +1,427 @@
+            contents.append(line);
+            contents.append(System.getProperty("line.separator"));
+          }
+        } catch (IOException e) {
+          e.printStackTrace();
+          Log.d(TAG, "Error reading line with BufferedReader");
+        } finally {
+          input.close();
+        }
+
+      } catch (IOException e) {
+        e.printStackTrace();
+        Log.d(TAG, "Error closing BufferedReader");
+      }
+
+      addNoteEditText.setText(contents.toString()); // Display the note content
+
+      enableShareNoteMenuItem(); // As note content exists which can be shared
+    }
+
+    // No action in case the note file for the currently open article doesn't exist
+  }
+
+  private void shareNote() {
+
+    /* The note text file corresponding to the currently open article, given at:
+     *    "{External Storage}/Kiwix/Notes/ZimFileTitle/ArticleTitle.txt"
+     * is shared via an app-chooser intent
+     * */
+
+    if(noteEdited) {
+     saveNote(addNoteEditText.getText().toString()); // Save edited note before sharing the text file
+    }
+
+    File noteFile = new File(Environment.getExternalStorageDirectory() + "/Kiwix/Notes/" + zimFileTitle + "/" + articleTitle + ".txt");
+
+    Uri noteFileUri = null;
+    if(noteFile.exists()) {
+
+      if (Build.VERSION.SDK_INT >= 24) {
+        // From Nougat 7 (API 24) access to files is shared temporarily with other apps
+        // Need to use FileProvider for the same
+        noteFileUri = FileProvider.getUriForFile(getContext(), BuildConfig.APPLICATION_ID+".fileprovider", noteFile);
+
+      } else {
+        noteFileUri = Uri.fromFile(noteFile);
+      }
+
+    } else {
+      Toast.makeText(getContext(), getString(R.string.note_share_error_file_missing), Toast.LENGTH_SHORT).show();
+    }
+
+    if(noteFileUri != null) {
+      Intent noteFileShareIntent = new Intent(Intent.ACTION_SEND);
+      noteFileShareIntent.setType("application/octet-stream");
+      noteFileShareIntent.putExtra(Intent.EXTRA_STREAM, noteFileUri);
+      noteFileShareIntent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
+      Intent shareChooser = Intent.createChooser(noteFileShareIntent, getString(R.string.note_share_app_chooser_title));
+
+      if(noteFileShareIntent.resolveActivity(getActivity().getPackageManager()) != null) {
+        startActivity(shareChooser);
+      }
+    }
+  }
+
+  public static boolean isExternalStorageWritable() {
+    String state = Environment.getExternalStorageState();
+    if(Environment.MEDIA_MOUNTED.equals(state)) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public void onStart() {
+    super.onStart();
+
+    Dialog dialog = getDialog();
+    if(dialog != null) {
+      int width = ViewGroup.LayoutParams.MATCH_PARENT;
+      int height = ViewGroup.LayoutParams.MATCH_PARENT;
+      dialog.getWindow().setLayout(width, height);
+    }
+  }
+
+}
+
+class ConfirmationAlertDialogFragment extends DialogFragment {
```

Comment:
Extract to separate file

Reply:
@macgills thanks for the tips & review!
Regarding this bit of code, this DialogFragment subclass is tailored for asking user confirmation in case of unsaved changes (in terms of the dialog message and functionality acc. user response), and so it wouldn't have any use in other activities.

Should I leave it like this, _OR_ create a generic DialogFragment super-class which
- takes in a dialogMessage string in its constructor, and
- uses a callback interface for responding to user clicks?

The latter could be used whenever a DialogFragment is to be displayed

Reply:
I would just extract to separate file and keep it as is for now but keep that idea in mind and if it comes up again reassess. 
A non static inner class is just always suspicious because they implicitly leak a reference from their containing class, this specific class explicitly takes a reference so not so much a major problem but inner classes are one of the worst culprits on android for creating context leaks.

Reply:
Refactored [here](https://github.com/kiwix/kiwix-android/pull/1198/commits/7b8bd2b406b4b842166cf0f3937294fa97d35e94)

I converted it into a generic class after all - I had to write a callback interface anyway, so making it generic felt more air-tight than leaving it as a helper class specifically for AddNoteDialog. It can now be used for displaying a 2-button (positive & negative) confirmation dialog fragment on top of an existing dialog fragment